### PR TITLE
feat(linter): add support for workspace rules

### PR DESCRIPTION
**Pull request by JamesHenry**
## Related Issue(s)

Fixes #

### CODE REVIEW

**Review by FrozenPandaz — CHANGES_REQUESTED** [review]
Webstorm doesn't seem to be picking up when I make changes to the local rule :thinking:

**Comment by JamesHenry** [comment]
Thanks @mattlewis92 it's a good call out, most workspaces will want to be able to test their lint rules programmatically as well as manually.

@FrozenPandaz and I discussed it and I have extended the PR to have the eslint-rules be a full project on the project-graph and generate a valid jest test target, as well as generate a spec for every rule that is generated.

**Comment by JamesHenry** [comment]
Build failing because of https://github.com/nrwl/nx/pull/6900, not related to this PR

**Comment by mattlewis92** [comment]
> Thanks @mattlewis92 it's a good call out, most workspaces will want to be able to test their lint rules programmatically as well as manually.
> 
> @FrozenPandaz and I discussed it and I have extended the PR to have the eslint-rules be a full project on the project-graph and generate a valid jest test target, as well as generate a spec for every rule that is generated.

Amazing, thank you so much!! 🙌

File: packages/linter/src/generators/workspace-rules-project/workspace-rules-project.ts [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+  readProjectConfiguration,
+  readWorkspaceConfiguration,
+  Tree,
+  updateWorkspaceConfiguration,
+} from '@nrwl/devkit';
+import { jestProjectGenerator } from '@nrwl/jest';
+import { join } from 'path';
+import { workspaceLintPluginDir } from '../../utils/workspace-lint-rules';
+
+export const WORKSPACE_RULES_PROJECT_NAME = 'workspace-lint-rules';
```
Comment (FrozenPandaz) [inline packages/linter/src/generators/workspace-rules-project/workspace-rules-project.ts:15]: I would call this `eslint-rules`.

**Review by FrozenPandaz — CHANGES_REQUESTED** [review]
Can we also add an `e2e` test for this please?

File: packages/linter/src/generators/workspace-rule/workspace-rule.ts [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+  name: string;
+  directory: string;
+}
+
+export async function lintWorkspaceRuleGenerator(
+  tree: Tree,
+  options: LintWorkspaceRuleGeneratorOptions
+) {
+  // Ensure that the workspace rules project has been created
+  await lintWorkspaceRulesProjectGenerator(tree);
```
Comment (FrozenPandaz) [inline packages/linter/src/generators/workspace-rule/workspace-rule.ts:27]: You need to return the task returned by this generator.

Currently, the npm install isn't triggered afterwards.

**Review by FrozenPandaz — CHANGES_REQUESTED** [review]

File: packages/linter/src/generators/workspace-rule/workspace-rule.ts (hunk 2 above)
Comment (JamesHenry) [reply packages/linter/src/generators/workspace-rule/workspace-rule.ts]: Thanks, done

**Comment by JamesHenry** [comment]
Thanks @FrozenPandaz, ready for another look

File: docs/map.json [hunk 3]
```diff
@@ -670,6 +670,11 @@
             "name": "lint executor",
             "id": "lint",
             "file": "angular/api-linter/executors/lint"
+          },
+          {
+            "name": "workspace-rule generator",
+            "id": "lint",
```
Comment (FrozenPandaz) [inline docs/map.json:676]: You need to change the ID of this to something unique. Currently, it is linking to the `lint` executor.

**Review by FrozenPandaz — CHANGES_REQUESTED** [review]

**Review by FrozenPandaz — APPROVED** [review]

**Pull request merged**